server/trillian_log_signer/main.go

```diff
@@ -0,0 +1,103 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"flag"
+	"fmt"
+	"net"
+	"net/http"
+	"os"
+	"os/signal"
+	"syscall"
+	"time"
+
+	"github.com/golang/glog"
+	"github.com/google/trillian/extension/builtin"
+	"github.com/google/trillian/server"
+	"github.com/google/trillian/util"
+	"golang.org/x/net/context"
+)
+
+var (
+	exportRPCMetrics              = flag.Bool("export_metrics", true, "If true starts HTTP server and exports stats")
+	httpPortFlag                  = flag.Int("http_port", 8091, "Port to serve HTTP metrics on")
+	sequencerSleepBetweenRunsFlag = flag.Duration("sequencer_sleep_between_runs", time.Second*10, "Time to pause after each sequencing pass through all logs")
+	batchSizeFlag                 = flag.Int("batch_size", 50, "Max number of leaves to process per batch")
+	numSeqFlag                    = flag.Int("num_sequencers", 10, "Number of sequencers to run in parallel")
+	sequencerGuardWindowFlag      = flag.Duration("sequencer_guard_window", 0, "If set, the time elapsed before submitted leaves are eligible for sequencing")
+)
+
+func startHTTPServer(port int) error {
+	sock, err := net.Listen("tcp", fmt.Sprintf("localhost:%d", port))
+	if err != nil {
+		return err
+	}
+	go func() {
+		glog.Info("HTTP server starting")
+		http.Serve(sock, nil)
+	}()
+
+	return nil
+}
+
+func awaitSignal(cancel context.CancelFunc) {
+	// Arrange notification for the standard set of signals used for termination.
+	sigs := make(chan os.Signal, 1)
+	signal.Notify(sigs, syscall.SIGINT, syscall.SIGTERM)
+
+	// Now block main and wait for a signal
+	sig := <-sigs
+	glog.Warningf("Signal received: %v", sig)
+	glog.Flush()
+
+	// Cancel the main context.
+	cancel()
+}
+
+func main() {
+	flag.Parse()
+	glog.CopyStandardLogTo("WARNING")
+	glog.Info("**** Log Signer Starting ****")
+
+	// First make sure we can access the database and keys, quit if not
+	registry, err := builtin.NewDefaultExtensionRegistry()
+	if err != nil {
+		glog.Exitf("Failed to create extension registry: %v", err)
+	}
+
+	// Start HTTP server (optional)
+	if *exportRPCMetrics {
+		glog.Infof("Creating HTP server starting on port: %d", *httpPortFlag)
+		if err := startHTTPServer(*httpPortFlag); err != nil {
+			glog.Exitf("Failed to start http server on port %d: %v", *httpPortFlag, err)
+		}
+	}
+
+	// Start the sequencing loop, which will run until we terminate the process. This controls
+	// both sequencing and signing.
+	// TODO(Martin2112): Should respect read only mode and the flags in tree control etc
+	ctx, cancel := context.WithCancel(context.Background())
+	go awaitSignal(cancel)
+
+	sequencerManager := server.NewSequencerManager(registry, *sequencerGuardWindowFlag)
+	sequencerTask := server.NewLogOperationManager(ctx, registry, *batchSizeFlag, *numSeqFlag, *sequencerSleepBetweenRunsFlag, util.SystemTimeSource{}, sequencerManager)
+	sequencerTask.OperationLoop()
+
```

Comment:
`cancel` is the only way to stop the `sequencerTask`, so to get here `cancel` must have been called (by the signal handler).